### PR TITLE
Enum parameter handling

### PR DESCRIPTION
**Pull request by jh66637**
As discussed in https://github.com/exadg/exadg/pull/264 I am filing a new PR for parameter handling with enums. Personally, I like the proposed interface, which automatically generates the ```dealii::Selections```. Thus, possible parameters can be printed, and it is checked if a given parameter is valid.

This PR is dependent on https://github.com/exadg/exadg/pull/264 and can only be merged after.

If a parameter is spelled wrong in, e.g., a `*.json` file the error Message shows which parameters can be used. For the test the output looks as follows:
```
38: set EnumType  =   Any one of Undefined, AScoped, BScoped, CScoped
38: 
38: # Listing of Parameters
38: # ---------------------
38: set EnumType = Undefined
38: 
38: terminate called after throwing an instance of 'dealii::ParameterHandler::ExcValueDoesNotMatchPattern'
38:   what():  
38: --------------------------------------------------------
38: An error occurred in line <1202> of file </home/johannes/workspace/dealii/source/base/parameter_handler.cc> in function
38:     void dealii::ParameterHandler::set(const string&, const string&)
38: The violated condition was: 
38:     patterns[pattern_index]->match(new_value)
38: Additional information: 
38:     The string <Undefineda> does not match the given pattern <[Selection
38:     Undefined|AScoped|BScoped|CScoped ]>.
```

### CODE REVIEW

**Review by peterrum — APPROVED** [review]

File: include/exadg/utilities/parameter_handler.h
Comment (jh66637) [reply include/exadg/utilities/parameter_handler.h]: I removed the namespace

**Review by peterrum — APPROVED** [review]
Good to go once you have rebased!

**Comment by jh66637** [comment]
@peterrum @nfehn Should this PR also contain the replacement of all places in which the enum parameter handling simplifies the code (where a string is given as an input and converted to an enum) or should this be done independently?

**Comment by jh66637** [comment]
@nfehn @richardschu I finally rebased. And I applied the parameter handler in various places. 
I think this is ready to merge and makes https://github.com/exadg/exadg/pull/367 obsolete :)

File: applications/incompressible_navier_stokes/flow_past_cylinder/application.h
Comment (nfehn) [inline applications/incompressible_navier_stokes/flow_past_cylinder/application.h:147]: clang-format is off here, so optimally we would care manually about nice indentation here.

File: applications/incompressible_navier_stokes/taylor_green_vortex/application.h
Comment (nfehn) [inline applications/incompressible_navier_stokes/taylor_green_vortex/application.h:105]: We would need to fix indentation manually here.

Should probably be `Curvilinear` (capital "C") to avoid misunderstandings?!

File: applications/structure/bar/application.h
Comment (nfehn) [inline applications/structure/bar/application.h:390]: Why is it just `Dirichlet` and not something like `BoundaryType::Dirichlet`?

File: applications/structure/bar/application.h
Comment (nfehn) [inline applications/structure/bar/application.h:526]: enum class?

File: applications/structure/can/application.h
Comment (nfehn) [inline applications/structure/can/application.h:299]: enum class?

File: include/exadg/incompressible_navier_stokes/driver.h
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/driver.h:77]: why is `true` removed here?

File: include/exadg/utilities/throughput_parameters.h
Comment (nfehn) [inline include/exadg/utilities/throughput_parameters.h:32]: not related to this PR, but should for consistency read `<..../print_solver_results.h>`

File: include/exadg/utilities/parameter_handler.h
Comment (nfehn) [inline include/exadg/utilities/parameter_handler.h:29]: I do not get the purpose of this file.

File: include/exadg/utilities/throughput_parameters.h
Comment (jh66637) [reply include/exadg/utilities/throughput_parameters.h:32]: I filed a new PR for this https://github.com/exadg/exadg/pull/466

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:45]: we typically write `typename` in ExaDG, and use `const` as trailing keyword (see functions below).

**Comment by nfehn** [comment]
As a general comment: I do not want to give up the parameter `has_to_be_set` in `ParameterHandler::add_parameter()`. This needs to be discussed before we can merge this PR.

**Comment by nfehn** [comment]
Overall, this PR is good work and I am very much looking forward to have this in ExaDG!

**Comment by nfehn** [comment]
Could you briefly explain what happens if I mis-spell an enum type in the json input file? Do I get concrete hints which parameter causes problems and which options are available for this parameter? Maybe you could demonstrate this for one of the examples you have changed in this PR.

File: applications/incompressible_navier_stokes/flow_past_cylinder/application.h
Comment (jh66637) [reply applications/incompressible_navier_stokes/flow_past_cylinder/application.h:147]: I would like to go this way instead: https://github.com/exadg/exadg/pull/467 :)
Is this OK for you?

File: include/exadg/utilities/parameter_handler.h
Comment (jh66637) [reply include/exadg/utilities/parameter_handler.h:29]: It is important to include the two headers in this exact order. To avoid mixing this up and also to be able to have a clean interface for the user, I added the file.

File: applications/incompressible_navier_stokes/taylor_green_vortex/application.h
Comment (jh66637) [reply applications/incompressible_navier_stokes/taylor_green_vortex/application.h]: [fe22560](https://github.com/exadg/exadg/pull/382/commits/fe22560ec03fbd3134afc96a469110348093d320)

File: applications/structure/bar/application.h
Comment (jh66637) [reply applications/structure/bar/application.h]: [fe22560](https://github.com/exadg/exadg/pull/382/commits/fe22560ec03fbd3134afc96a469110348093d320)

File: applications/structure/bar/application.h
Comment (jh66637) [reply applications/structure/bar/application.h]: [fe22560](https://github.com/exadg/exadg/pull/382/commits/fe22560ec03fbd3134afc96a469110348093d320)

File: include/exadg/utilities/parameter_handler.h
Comment (nfehn) [reply include/exadg/utilities/parameter_handler.h:29]: It would make more sense in my opinion to include `dealii/base/parameter_handler.h` (with a suitable comment) in `<exadg/utilities/enum_patterns.h>`. The current choice with a new (empty) header file looks like needless complexity to me and is very hard to understand in my opinion.

**Comment by jh66637** [comment]
> As a general comment: I do not want to give up the parameter `has_to_be_set` in `ParameterHandler::add_parameter()`. This needs to be discussed before we can merge this PR.

I added a utility function (see [5ccc93b](https://github.com/exadg/exadg/pull/382/commits/5ccc93bc65e66fef51f7432d8b4af059944a3426)) to make it easy to set `has_to_be_set` again :)

**Comment by jh66637** [comment]
> Could you briefly explain what happens if I mis-spell an enum type in the json input file? Do I get concrete hints which parameter causes problems and which options are available for this parameter? Maybe you could demonstrate this for one of the examples you have changed in this PR. You could add this to the description of this PR (also for the purpose of documentation of the changes/functionality introduced here).

Yes. The allowed pattern is generated automatically using the utility function to serialize an `enum` which has been introduced before :) I will add it to the discription :)

File: include/exadg/incompressible_navier_stokes/driver.h
Comment (jh66637) [reply include/exadg/incompressible_navier_stokes/driver.h:77]: fixed, as discussed below :)

**Comment by peterrum** [comment]
@jh66637 May I ask you to add a test? You can simply adopt mine: https://github.com/peterrum/dealii-parameter-handler-enum/tree/master/tests

File: applications/incompressible_navier_stokes/flow_past_cylinder/application.h
Comment (jh66637) [reply applications/incompressible_navier_stokes/flow_past_cylinder/application.h]: I fixed this indentation in [7fd5759](https://github.com/exadg/exadg/pull/382/commits/7fd5759d839140a3d130e746455381be06bce9ed)

File: include/exadg/utilities/parameter_handler.h
Comment (jh66637) [reply include/exadg/utilities/parameter_handler.h:29]: If it is OK for you to include `#include <exadg/utilities/enum_patterns.h>` instead of a `parameter_handler.h` its OK for me.

Comment (nfehn) [reply include/exadg/utilities/parameter_handler.h:29]: You mean including dealii/parameter_handler additionally in other files in ExaDG will give an error?

File: include/exadg/convection_diffusion/throughput.h
Comment (nfehn) [inline include/exadg/convection_diffusion/throughput.h:32]: Is this removal mandatory?

File: include/exadg/utilities/parameter_handler.h
Comment (nfehn) [reply include/exadg/utilities/parameter_handler.h:29]: see comment https://github.com/exadg/exadg/pull/382/files#r1214137407

Comment (jh66637) [reply include/exadg/utilities/parameter_handler.h:29]: No, it is still possible to include parameter handler. But if one wants to treat `enums` one has to include `enum_patterns.h` in addition. I think I will adjust it this way.

**Comment by jh66637** [comment]
> @jh66637 May I ask you to add a test? You can simply adopt mine: https://github.com/peterrum/dealii-parameter-handler-enum/tree/master/tests

@peterrum Thanks for pointing to your tests: I added a similar test :)

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: I extended the documentation in [c3f0a7d](https://github.com/exadg/exadg/pull/382/commits/c3f0a7daf7857d39dcd9474eb86602d5bde96c1f) as suggested. I still want to cite the project since it helped a lot during development of the code.

File: include/exadg/utilities/parameter_handler.h
Comment (jh66637) [reply include/exadg/utilities/parameter_handler.h]: I changed it in [c3f0a7d](https://github.com/exadg/exadg/pull/382/commits/c3f0a7daf7857d39dcd9474eb86602d5bde96c1f)

File: include/exadg/convection_diffusion/throughput.h
Comment (jh66637) [reply include/exadg/convection_diffusion/throughput.h:32]: No, see [c3f0a7d](https://github.com/exadg/exadg/pull/382/commits/c3f0a7daf7857d39dcd9474eb86602d5bde96c1f)

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:42]: dunctions->functions

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:40]: Unclear formulation in my opinion: "possible parameters" -> "the allowed values of a parameter"?

What does "is also possible" mean exactly?

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:91]: I would suggest `Patterns::Enum<...>` to be closer to the deal.II syntax.

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:93]: Why can't we place the dereferencing `*` here?

File: include/exadg/utilities/hypercube_resolution_parameters.h
Comment (nfehn) [inline include/exadg/utilities/hypercube_resolution_parameters.h:181]: see comment https://github.com/exadg/exadg/pull/382/files#r1214261462

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [inline include/exadg/utilities/enum_patterns.h:88]: Does "//" invoke documentation via Doxygen?

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: Is this more clear? [20dc382](https://github.com/exadg/exadg/pull/382/commits/20dc38232130347f34f598e32c09320d27eff01a)

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: [20dc382](https://github.com/exadg/exadg/pull/382/commits/20dc38232130347f34f598e32c09320d27eff01a)

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: It's an abstract base class, so `dealii::Patterns::PatternBase` as a return type will not work. I am now returning `dealii::Patterns::PatternBase const&` instead, which makes it possible to shift the `*` inside the utility function. [220423d](https://github.com/exadg/exadg/pull/382/commits/220423d5e62e7a26203e30351f741d8f7a1f2beb)

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: [220423d](https://github.com/exadg/exadg/pull/382/commits/220423d5e62e7a26203e30351f741d8f7a1f2beb)

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [reply include/exadg/utilities/enum_patterns.h]: This is good! 

I think one would need a "," after the if clause (before deal.II).

File: include/exadg/utilities/hypercube_resolution_parameters.h
Comment (jh66637) [reply include/exadg/utilities/hypercube_resolution_parameters.h]: [220423d](https://github.com/exadg/exadg/pull/382/commits/220423d5e62e7a26203e30351f741d8f7a1f2beb)

File: include/exadg/utilities/enum_patterns.h
Comment (nfehn) [reply include/exadg/utilities/enum_patterns.h]: Yes, this was exactly my intention. Great!

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: [4e0a0ff](https://github.com/exadg/exadg/pull/382/commits/4e0a0ffb39ac982da6988f8947f6f2019f6f5ab4)

File: include/exadg/utilities/enum_patterns.h
Comment (jh66637) [reply include/exadg/utilities/enum_patterns.h]: Done :)

**Review by nfehn — APPROVED** [review]

**Comment by nfehn** [comment]
Very nice result!

**Comment by jh66637** [comment]
Thank you for the valuable review. I think I applied all requested changes. This is ready from my side once the tests passed :)

File: tests/utilities/enum_parameter_handler.output
Comment (peterrum) [inline tests/utilities/enum_parameter_handler.output:27]: Thanks!

**Comment by nfehn** [comment]
Due to the new namespace `ExaDG::Patterns`, one probably now has to write explicitly `dealii::Patterns::Double`.

I wonder whether we still have `using namespace dealii` somewhere in the code ...

**Comment by jh66637** [comment]
I fixed the tests (at least at my local machine). Handing back a const reference of an abstract base class is obviousely problematic even though it compiles ^^

**Comment by jh66637** [comment]
@nfehn Finally all tests (except parabolic_inflow) passed :)

**Comment by nfehn** [comment]
Could you point me to the code that has been adjusted to fix the tests?

**Comment by jh66637** [comment]
Instead of handing back the underlying `Pattern` base class, I am now returning `Pattern::Selection`.
See:
https://github.com/exadg/exadg/pull/382/files#diff-e62885cbf25268722e055c1c4668f964d14bb653a7028f45d514a6f1deed0648R55-R59
and
https://github.com/exadg/exadg/pull/382/files#diff-e62885cbf25268722e055c1c4668f964d14bb653a7028f45d514a6f1deed0648R94-R99

**Pull request merged**
